Improve web acceptance contribution doc (#505)
PrestaShop/docs: src/content/1.7/testing/web-acceptance-tests/how-to-contribute-and-create-web-acceptance-tests.md

---
title: How to contribute and create web acceptance tests

> menuTitle: Contributing to the creation of web acceptance tests

Reviewer: This is maybe a little long for the menu title?

> weight: 2

Reviewer: You need to create an alias for the previous route you deleted, or else there will be broken links and lost page positioning

```suggestion
weight: 2
aliases:
   - /1.7/testing/how-to-create-your-own-web-acceptance-tests
```

---

> ## How to contribute and create web acceptance tests

Reviewer: Your header hierarchy is wrong, this should be an H1, the following header an H2, and so on

```suggestion
# How to contribute and create web acceptance tests
```

When you create new page classes, make sure to make them inherit from their respective `XXMainPage` to be sure to be able to use both generic methods and specific methods.

Examples: 

- Install page (which is independent) extends from `CommonPage`
- Brands page (which is part of the BO) extends from `BOBasePage`
- Checkout page (which is part of the FO) extends from `FOBasePage`  

#### Methods
Methods inside a page class must be properly named about their primary function. Remember to [KISS](https://en.wikipedia.org/wiki/KISS_principle) ! You should be able to understand what a method does just by looking at its name.

Methods inside a page class **CANNOT USE** the `expect` keyword or any kind of asserting function, since their sole purpose is to interact with the page they’re linked with. They don’t validate anything per se, they just expose functionalities. 
To validate a behavior in your test logic, your methods can return booleans, integers (number of lines, elements, etc), strings, objects... It’s up to you.

Examples: 

> ```

Reviewer: ```suggestion
```js
```

Author: This how it display **with** js 
![image](https://user-images.githubusercontent.com/48441421/80367595-aea37100-888b-11ea-9482-5f5a91611c47.png)

And this is how it display **without** js
![image](https://user-images.githubusercontent.com/48441421/80367709-dbf01f00-888b-11ea-8997-c210aac670f5.png)

For better visibility, i choose to not use it, WDYT ?

Reviewer: That's quite odd... see, we use `js` already in some pages like [this one](https://github.com/PrestaShop/docs/blob/master/src/content/1.7/development/components/grid/columns-reference/position.md#L46) and it [renders well](https://devdocs.prestashop.com/1.7/development/components/grid/columns-reference/position/):

<img width="763" alt="Screenshot 2020-04-30 at 17 42 41" src="https://user-images.githubusercontent.com/1009343/80730559-051de300-8b0a-11ea-9835-b001c5128ddb.png">

Author: I think it's broken when testing in local machine, so I added js

   /**
   * Get order status
   * @return {Promise<string>}
   */
  async getOrderStatus() {
    return this.getTextContent(`${this.orderStatusesSelect} option[selected='selected']`, false);
  }
```

#### Selectors
Selectors are used on every page and are stored as attributes of the class. They should be named following this convention:

**nameType*** in `camelCase`.

For example, a button used to submit the main form in the order page should be named: `submitMainFormButton`.

#### Types
Each selector must belong to a certain type. Here is a non-exhaustive list:

> - button

Reviewer: You need to separate lists from the previous paragraph or else it won't work properly.

![Screenshot 2020-04-27 at 13 16 52](https://user-images.githubusercontent.com/1009343/80366321-63885e80-8889-11ea-9672-c84ff524d5b3.png)

I **strongly** suggest previewing your pages locally with Hugo (it's [very easy](https://github.com/PrestaShop/docs#rendering-the-site-locally)) before submitting them.

```suggestion

- button
```

- link
- block
- image
- icon
- text
- modal
- *other HTML elements...*

- table
- table-header
- table-cell
- table-row
- *other table elements...*

- form
- input
- select
- radio
- checkbox
- *other forms inputs...*

### Tests

#### Campaigns
We currently have 2 campaigns implemented:

> - **Sanity**: its purpose is to validate a Pull Request. Executed for every one on [Travis CI](https://travis-ci.com/), this campaign must fully pass before merging the PR (one failed test blocks the merge). It consists of a few tests of the core features of PrestaShop: shop installation, orders/products pages in BO, and catalog/cart/checkout process in FO.

Reviewer 2: ```suggestion
- **Sanity**: its purpose is to validate a Pull Request. Executed every time on [Travis CI](https://travis-ci.com/), this campaign must fully pass before merging the PR (one failed test blocks the merge). It consists of a few tests of the core features of PrestaShop: shop installation, orders/products pages in BO, and catalog/cart/checkout process in FO.
```

Author: ```suggestion
- **Sanity**: its purpose is to validate a Pull Request. Executed on [Travis CI](https://travis-ci.com/), this campaign must fully pass before merging the PR (one failed test blocks the merge). It consists of a few tests of the core features of PrestaShop: shop installation, orders/products pages in BO, and catalog/cart/checkout process in FO.
```

better maybe ?

Reviewer 2: Yes !

- **Functional**: it is the biggest and most important campaign. Its purpose is to validate that every feature of PrestaShop works, by testing them one by one. It goes on every page and tests whatever it can: table (filtering, ordering, quick edits, etc), [CRUD](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete) items (orders, customers, products…), setting changes, etc.

We plan on implementing 2 more campaigns:

> - **End to end**: its purpose is to check that popular user paths are working as intended. It will walk through the application and mimic a real user working on their daily routines as a merchant: checking products, generating invoices, creating a customer account, and an order, adding a special voucher for a specific user, etc. The selected user paths will be chosen by the Product Team. Thanks to their merchants and agencies interviews, they have a pretty good idea of what merchants do every day and how they use the software.

Reviewer: Again, separate the first list item from the previous paragraph.
```suggestion

- **End to end**: its purpose is to check that popular user paths are working as intended. It will walk through the application and mimic a real user working on their daily routines as a merchant: checking products, generating invoices, creating a customer account, and an order, adding a special voucher for a specific user, etc. The selected user paths will be chosen by the Product Team. Thanks to their merchants and agencies interviews, they have a pretty good idea of what merchants do every day and how they use the software.
```

- **Regression**: a test campaign tailored to only target major and critical issues in the few last versions of PrestaShop, to make sure they don’t come back (hence the name).

#### Mocha and Mochawesome
[Mocha](https://mochajs.org/) is our test runner (a framework that reads our test code and runs it). We use it in coordination with [Mochawesome](https://www.npmjs.com/package/mochawesome), a plugin for Mocha. Mochawesome produces a full JSON report in addition to a beautiful HTML report (which we don’t really use). The JSON file is sent to our [Nightly Board](https://nightly.prestashop.com/), it is then inserted into the database to let visitors browse reports and visualize statistics.

> #### Lambda function (Why ?)

Reviewer: ```suggestion
#### Lambda function (Why?)
```

The functions that exist (for now) in this file are the following : 
- `createBrowser`: used to create a browser with the global configuration
- `closeBrowser`: usually called at the end of a test, to close the browser created for the test
- `newTab`: allow us to open a new tab in a browser
- `setDownloadBehavior`: used only for tests that contain files downloading

Note that all these functions are used at mocha hooks functions in the global `describe` but can be called somewhere else.

##### Files
Some of our tests need to create files (ex: Create files in BO), or to check some text in a PDF file (ex: Create and check invoice). For this specific need, we use functions in `Files.js`.

When a test is finished, all created/downloaded files are deleted using a function from the same file : `deleteFile` as part of the "cleaning behind" approach.

#### Initialize pages
In each and every test, we initialize the pages that will be needed. The initialization is done in a function called `init` which returns an object with multiple entries (pages) of pages.

This function is called at the beginning of the test and each time we change the browser’s tab.

Example: 

> ```

Reviewer: ```suggestion
```js
```

// For test ‘Filter Customer’
const init = async function () {
  return {
    loginPage: new LoginPage(page),
    dashboardPage: new DashboardPage(page),
    customersPage: new CustomersPage(page),
  };
};
```
  
#### Expect
We use the `expect` keyword from the [Chai library](https://www.chaijs.com/api/bdd/). This allows us to write assertions in a much more readable way. You can use whatever way to assert you want/need. Don’t forget that you can use the second argument of `expect` to log out a better error message when your assertion fails.

Example :

> ``` 

Reviewer: ```suggestion
```js
```

Const isCustomerConnected = await this.pageObject.foLoginPage.isCustomerConnected();
await expect(isCustomerConnected, ‘Customer is disconnected in FO’).to.be.true;
```  

#### Test identifier
Our team thinks it’s very important to be able to follow how tests results evolve, so it’s been decided to add a unique identifier to every step in the test.

Why? To be able to identify each and every test and compare from one report to another, to know how the results change.

For example, you could have 10 failed tests one day, and 10 the day after. How do you know if it’s the same 10 tests failing, or another distribution (for example, 5 tests fixed, 5 other tests failing)? With this system we can calculate the evolution and show it on the Board.

We first create a base context for each and every test file, and then we make a call to the function `addContextItem` with the unique value inside the test file. 

Example :

> ```

Reviewer: ```suggestion
```js
```

#### Types
Each selector must belong to a certain type. The whole list (not exhaustive) is here:
- button
- link
- block
- image
- icon
- text
- modal
- *other HTML elements...*

- table
- table-header
- table-cell
- table-row
- *other table elements...*

- form
- input
- select
- radio
- checkbox
- *other forms inputs...*

### Tests

#### Campaigns
We currently have 2 campaigns implemented:

- **Sanity**: its purpose is to validate a Pull Request. It is executed in a CI environment (on Travis CI) for every Pull Request and is mandatory, i.e. a failed test blocks the merge. It consists of a few tests of the core features of PrestaShop: orders in BO, Products in BO, catalog in FO, cart in FO, Checkout process in FO, and installation of the build.
- **Functional**: it’s the biggest and most important campaign. Its purpose is to validate that every feature of PrestaShop works, by testing them one by one. It goes on every page and tests whatever it can: table (filtering, ordering, quick edits, etc), CRUD (Create Read Update Delete) of items (Orders, Customers, Products, and so on), settings changes…

We plan on implementing 2 more campaigns:

> - **End to end**: its purpose is to check that popular user paths are working as intended. It will walk through the application and mimic a real user working on their daily routines as a merchant: checking products, generating invoices, creating a customer account and an order, adding a special voucher for a specific user, etc. The selected user paths will be chosen by the Product Team. Thanks to their merchants and agencies interviews, they have a pretty good idea of what a merchant does every day and how they use the software.

Reviewer 2: ```suggestion
- **End to end**: its purpose is to check that popular user paths are working as intended. It will walk through the application and mimic a real user working on their daily routines as a merchant: checking products, generating invoices, creating a customer account and an order, adding a special voucher for a specific user, etc. The selected user paths will be chosen by the Product Team. Thanks to their merchants and agencies interviews, they have a pretty good idea of what merchants do every day and how they use the software.
```

> - **Regression**: a test campaign tailored to only target majors and critical issues in the few last versions of PrestaShop, to make sure they don’t come back (hence the name).

Reviewer 2: ```suggestion
- **Regression**: a test campaign tailored to only target major and critical issues in the few last versions of PrestaShop, to make sure they don’t come back (hence the name).
```

#### Mocha and Mochawesome

> Mocha (https://mochajs.org/) is our test runner (a framework that reads our test code and runs it). We use it in coordination with Mochawesome (https://www.npmjs.com/package/mochawesome), a plugin for Mocha. Mochawesome produces a full JSON report in addition to a beautiful HTML report (which we don’t really use). The JSON file is sent to our Nightly Board, it is then inserted into the database to let visitors browse reports and visualize statistics.

Reviewer 2: ```suggestion
[Mocha](https://mochajs.org/) is our test runner (a framework that reads our test code and runs it). We use it in coordination with [Mochawesome](https://www.npmjs.com/package/mochawesome), a plugin for Mocha. Mochawesome produces a full JSON report in addition to a beautiful HTML report (which we don’t really use). The JSON file is sent to our [Nightly Board](https://nightly.prestashop.com/), it is then inserted into the database to let visitors browse reports and visualize statistics.
```

#### Lambda function (Why ?)

> Using lambda functions in mocha is discouraged (https://mochajs.org/#arrow-functions). They bind `this` to the scope of the lambda function, making it impossible to use internal Mocha methods and objects. Since we use the Mocha context to store our Page Objects, we strongly advise you to use the normal `function()` syntax. 

Reviewer 2: ```suggestion
Using lambda functions in mocha is [discouraged](https://mochajs.org/#arrow-functions). They bind `this` to the scope of the lambda function, making it impossible to use internal Mocha methods and objects. Since we use the Mocha context to store our Page Objects, we strongly advise you to use the normal `function()` syntax. 
```

#### Utils

> The utils directory contain files that are necessary to run tests

Reviewer 2: ```suggestion
The utils directory contain files that are necessary to run tests.
```

##### Globals
This file contains all global variables that can be used in test files, pages and common tests.

> The description of each variable in this file can be found in README.md.

Reviewer 2: ```suggestion
The description of each variable in this file can be found in [README.md](https://github.com/SimonGrn/PrestaShop/blob/develop/tests/puppeteer/README.md).
```

##### Browser helper

> The helper is used to centralize browser and tab functions called in all tests. 

Reviewer 2: ```suggestion
This helper class is used to centralize browser and tab functions called in all tests. 
```

Author: changed by `This helper file`, since it's not a class

This approach has one goal : to have the same browser’s configuration everywhere and of course to facilitate maintenance.

The functions that exist (for now) in this file are the following : 
- `createBrowser` : used to create a browser with the global configuration
- `closeBrowser` : usually called at the end of a test, to close the browser created for the test
- `newTab` : allow us to open a new tab in a browser

> - `setDownloadBehavior` : employed only for tests that contain files downloading

Reviewer 2: ```suggestion
- `setDownloadBehavior` : used only for tests that contain files downloading
```

> Note that all these functions are used at mocha hooks functions in the global describe but can be called somewhere else.

Reviewer 2: ```suggestion
Note that all these functions are used at mocha hooks functions in the global `describe` but can be called somewhere else.
```

##### Files
Some of our tests need to create files (ex: Create file in BO), or to check some text in a PDF file (ex: Create and check invoice). For this specific need, we use functions in `Files.js`.

> When a test is finished, all created/downloaded files are deleted using a function from the same file : ‘deleteFile` as part of the cleaning behind approach.

Reviewer 2: ```suggestion
When a test is finished, all created/downloaded files are deleted using a function from the same file : `deleteFile` as part of the "cleaning behind" approach.
```

#### Initialize pages
In each and every test, we initialize the pages that will be needed. The initialization is done in a function called `init` which returns an object with multiple entries (pages) of pages.

This function is called at the beginning of the test and each time we change the browser’s tab.

Example: 
```
// For test ‘Filter Customer’
const init = async function () {
  return {
    loginPage: new LoginPage(page),
    dashboardPage: new DashboardPage(page),
    customersPage: new CustomersPage(page),
  };
};
```
  
#### Expect

> We use the `expect` keyword from the `chai` library (https://www.chaijs.com/api/bdd/). This allows us to write assertions in a much more readable way. You can use whatever way to assert you want/need. Don’t forget that you can use the second argument of `expect` to log out a better error message when your assertion fails.

Reviewer 2: ```suggestion
We use the `expect` keyword from the [Chai library](https://www.chaijs.com/api/bdd/). This allows us to write assertions in a much more readable way. You can use whatever way to assert you want/need. Don’t forget that you can use the second argument of `expect` to log out a better error message when your assertion fails.
```

Example :

``` 
Const isCustomerConnected = await this.pageObject.foLoginPage.isCustomerConnected();
await expect(isCustomerConnected, ‘Customer is disconnected in FO’).to.be.true;
```  

#### Test identifier

> Our team thinks it’s very important to be able to follow how tests results evolve. So it’s been decided to add a unique identifier to every step in the test.

Reviewer 2: ```suggestion
Our team thinks it’s very important to be able to follow how tests results evolve, so it’s been decided to add a unique identifier to every step in the test.
```

Why? To be able to identify each and every test and compare from one report to another, to know how the results change.

For example, you could have 10 failed tests one day, and 10 the day after. How do you know if it’s the same 10 tests failing, or another distribution (for example, 5 tests fixed, 5 other tests failing)? With this system we can calculate the evolution and show it on the Board.

> For that, we first created a base context for each and every test file, and after that we make a call to the function `addContextItem` with the unique value inside the test file. 

Reviewer 2: ```suggestion
We first create a base context for each and every test file, and then we make a call to the function `addContextItem` with the unique value inside the test file. 
```

Example :

```
// From test : puppeteer/campaigns/functional/BO/04_customers/01_customers/07_helpCard.js
const baseContext = 'functional_BO_customers_customers_helpCard';
// And inside each `it`, we make a call
// For example, in the “Go To Customer’s Page” step we will have :
testContext.addContextItem(this, 'testIdentifier', 'goToCustomersPage', baseContext);
// In the report, the final identifier will look like this: ‘functional_BO_customers_customers_filterAndQuickEditCustomers_goToCustomersPage’
```

Be careful, identifiers **must be unique** through the whole campaign !

#### Cleaning behind
You must be able to launch a test independently, as well as in a whole campaign. That means :

> - Your test must create its own data or rely on the default fixtures
> - Your test must clean behind itself in a reliable manner
> - Deleting files (invoices, images) and artifacts

Reviewer 2: ```suggestion
* Your test must create its own data or rely on the default fixtures
* Your test must clean behind itself in a reliable manner
* Deleting files (invoices, images) and artifacts
```

Author: I don't understand this suggestion, it displays like this 
![image](https://user-images.githubusercontent.com/48441421/79586996-2b875b80-80d2-11ea-8d7e-6e199c32c2dd.png)

The shop must end in the same state it was in before your test, as much as possible (since some actions are logged and create artefacts, that may not be always easy to clean though) and let subsequent tests run smoothly! That means deleting the items you created, reverting your changes, etc.

A rule of thumb: can you launch your test suite multiple times? If yes, you know you’re not dependent on the data and you’re properly cleaning behind.

### Data

#### Demo Data
Our tests rely heavily on the demo fixtures (= demo data added when installing the vanilla PrestaShop package). However, we describe these data in separate files to make sure there’s no reference hard written in our code.

The only assumption we have to make is the presence of certain items like Orders or Products in the catalog.

If you need to rely on the fixtures too, make sure to use the description of the objects you’re looking for in the `data` folder. If it’s not complete, you can expand it and make a Pull Request, we’ll be happy to improve our datasets !

#### Faker

> When we need to create new items, we rely on Faker(https://github.com/marak/Faker.js/) to create random data. 

Reviewer 2: ```suggestion
When we need to create new items, we rely on [Faker](https://github.com/marak/Faker.js/) to create random data. 
```

### Lambda function (Why?)
Using lambda functions in mocha is [discouraged](https://mochajs.org/#arrow-functions). They bind `this` to the scope of the lambda function, making it impossible to use internal Mocha methods and objects. Since we use the Mocha context to store our Page Objects, we strongly advise you to use the normal `function()` syntax.  

### Utils
The utils directory contain files that are necessary to run tests.

#### Globals
This file contains all global variables that can be used in test files, pages and common tests.

The description of each variable in this file can be found in [README.md](https://github.com/PrestaShop/PrestaShop/blob/develop/tests/puppeteer/README.md).

#### Browser helper
This helper file is used to centralize the browser and tab functions called in all tests. 
This approach has one goal : to have the same browser’s configuration everywhere and of course to facilitate maintenance.

> The functions that exist (for now) in this file are the following :

Reviewer: ```suggestion
The functions that exist (for now) in this file are the following:
```
